### PR TITLE
Make Optional<bool> trivially copyable

### PR DESCRIPTION
**Pull request by rousskov**
ipc/TypedMsgHdr.h: static assertion failed: putPod() used for a POD
    ActionParams.cc:44: required from here [with Pod = RequestFlags]

The known XXX in Optional destructor has started to bite us because
pending changes expose memcpy(3)-based IPC serialization code to
Optional<bool> flags. It is possible to mimic standard std::optional
implementations, avoiding that XXX, but that requires rather
sophisticated C++ tricks with placement new() and such. Specializing the
whole Optional is a better alternative for this _temporary_ class IMO.

### CODE REVIEW

File: src/base/Optional.h [hunk 1]
```diff
@@ -122,6 +122,53 @@ class Optional
...
+    constexpr Value value_or(Other &&defaultValue) const &
+    {
+        return hasValue_ ? value_ : static_cast<Value>(std::forward<Other>(defaultValue));
+    }
+
+    template <class Other = Value>
+    Optional &operator =(Other &&otherValue)
+    {
+        value_ = std::forward<Other>(otherValue);
+        hasValue_ = true;
```
[the remark this answers is not in the document]
Comment (rousskov) [reply src/base/Optional.h:157]: > This looks like invalid move semantics

It does look invalid, but it is the correct semantics in this case. See generic Optional docs for this method:

```C++
Optional(Optional<Value> &&other): Optional()
{
    if (other.hasValue_) {
        *this = std::move(other.value_);
        // no other.reset() per std::optional move semantics
    }
}
```

From https://en.cppreference.com/w/cpp/utility/optional/optional:

> 3) Move constructor: If other contains a value, initializes the contained value as if [direct-initializing](https://en.cppreference.com/w/cpp/language/direct_initialization) (but not direct-list-initializing) an object of type T with the expression std::move(*other) and does not make other empty: a moved-from optional still contains a value, but the value itself is moved from.

[the remark this answers is not in the document]
Comment (rousskov) [reply src/base/Optional.h:157]: :exclamation: My earlier [response](https://github.com/squid-cache/squid/pull/1188/files#r1035003281) more-or-less applies to the change request [suggestion](https://github.com/squid-cache/squid/pull/1188/files#r1034914264), but not to the PR code in question: This change request suggestion implied that the change request is about a standard move assignment operator. My earlier response discussed a standard move _constructor_ instead (i.e. item 3 on cppreference as quoted in my earlier response). Sorry. However, the standard move assignment follows exactly the same invalid-looking logic (see item 3 on the corresponding [cppreference](https://en.cppreference.com/w/cpp/utility/optional/operator%3D) page):

> 3). Note that a moved-from optional still contains a value. 

The PR _code_ in question is _not_ a standard move assignment operator (i.e. not item 3)! This operator works on a value object rather than on an Optional<Value> object. It is actually a value conversion/moving assignment operator (for the lack of a better term). Here, we are creating an Optional<Value> object from a Value-compatible value object. That is why hasValue_ is set to true unconditionally -- we _were_ given a value. This operator is described in item 4 on [cppreference](https://en.cppreference.com/w/cpp/utility/optional/operator%3D) page for `std::optional::operator =`.

In summary, there are no known problems in this PR code, but this thread can be used as one more illustration of how complex and tricky std::optional details can be!
